engine/resources/docker_container.go

```diff
@@ -0,0 +1,441 @@
+	}
+
+	if len(containerList) > 1 {
+		return false, fmt.Errorf("more than one container named %s", obj.Name())
+	}
+	if len(containerList) == 0 && obj.State == ContainerRemoved {
+		return true, nil
+	}
+	if len(containerList) == 1 {
+		// If the state and image are correct, we're done.
+		if containerList[0].State == obj.State && containerList[0].Image == obj.Image {
+			return true, nil
+		}
+		id = containerList[0].ID // save the id for later
+		// If the image is wrong, and force is true, mark the container for
+		// destruction.
+		if containerList[0].Image != obj.Image && obj.Force {
+			destroy = true
+		}
+		// Otherwise return an error.
+		if containerList[0].Image != obj.Image && !obj.Force {
+			return false, fmt.Errorf("%s exists but has the wrong image: %s", obj.Name(), containerList[0].Image)
+
+		}
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	if obj.State == ContainerStopped { // container exists and should be stopped
+		return false, containerStop(ctx, obj.client, id, nil)
+	}
+
+	if obj.State == ContainerRemoved { // container exists and should be removed
+		if err := containerStop(ctx, obj.client, id, nil); err != nil {
+			return false, err
+		}
+		return false, containerRemove(ctx, obj.client, id, types.ContainerRemoveOptions{})
+	}
+
+	if destroy {
+		if err := containerStop(ctx, obj.client, id, nil); err != nil {
+			return false, err
+		}
+		if err := containerRemove(ctx, obj.client, id, types.ContainerRemoveOptions{}); err != nil {
+			return false, err
+		}
+		containerList = []types.Container{} // zero the list
+	}
+
+	if len(containerList) == 0 { // no container was found
+		// Download the specified image if it doesn't exist locally.
+		p, err := obj.client.ImagePull(ctx, obj.Image, types.ImagePullOptions{})
+		if err != nil {
+			return false, fmt.Errorf("error pulling image: %s", err)
+		}
+		// Wait for the image to download, EOF signals that it's done.
+		if _, err := ioutil.ReadAll(p); err != nil {
+			return false, fmt.Errorf("error reading image pull result: %s", err)
+		}
+
+		// set up port bindings
+		containerConfig := &container.Config{
+			Image:        obj.Image,
+			Cmd:          obj.Cmd,
+			Env:          obj.Env,
+			ExposedPorts: make(map[nat.Port]struct{}),
+		}
+
+		hostConfig := &container.HostConfig{
+			PortBindings: make(map[nat.Port][]nat.PortBinding),
+		}
+
+		for k, v := range obj.Ports {
+			for p, q := range v {
+				containerConfig.ExposedPorts[nat.Port(k)] = struct{}{}
+				hostConfig.PortBindings[nat.Port(fmt.Sprintf("%d/%s", p, k))] = []nat.PortBinding{
+					{
+						HostIP:   "0.0.0.0",
+						HostPort: fmt.Sprintf("%d", q),
+					},
+				}
+			}
+		}
+
+		c, err := obj.client.ContainerCreate(ctx, containerConfig, hostConfig, nil, obj.Name())
+		if err != nil {
+			return false, fmt.Errorf("error creating container: %s", err)
+		}
+		id = c.ID
+	}
+
+	return false, containerStart(ctx, obj.client, id, types.ContainerStartOptions{})
```

Comment:
why do you pass around obj.client instead of just having everything be a private struct method and using it directly??

Reply:
It made the tests a little easier to write, but I don't mind changing it if you don't like it.

Reply:
Ah. Interesting point. I think you can just init the tests with &DockerContainerRes{client: whatever,} and then call the methods like that.

Reply:
Yup, will do.